launcher_go/v2/config/config.go

```diff
@@ -0,0 +1,209 @@
+	Labels          map[string]string `yaml:labels,omitempty`
+	Volumes         []struct {
+		Volume struct {
+			Host  string `yaml:host`
+			Guest string `yaml:guest`
+		} `yaml:volume`
+	} `yaml:volumes,omitempty`
+	Links []struct {
+		Link struct {
+			Name  string `yaml:name`
+			Alias string `yaml:alias`
+		} `yaml:link`
+	} `yaml:links,omitempty`
+}
+
+func (config *Config) loadTemplate(templateDir string, template string) error {
+	template_filename := strings.TrimRight(templateDir, "/") + "/" + string(template)
+	content, err := os.ReadFile(template_filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			fmt.Println("template file does not exist: " + template_filename)
+		}
+		return err
+	}
+	templateConfig := &Config{}
+	if err := yaml.Unmarshal(content, templateConfig); err != nil {
+		return err
+	}
+	if err := mergo.Merge(config, templateConfig, mergo.WithOverride); err != nil {
+		return err
+	}
+	config.rawYaml = append(config.rawYaml, string(content[:]))
+	return nil
+}
+
+func LoadConfig(dir string, configName string, includeTemplates bool, templatesDir string) (*Config, error) {
+	config := &Config{
+		Name:         configName,
+		Boot_Command: DefaultBootCommand,
+		Base_Image:   DefaultBaseImage(),
+	}
+	matched, _ := regexp.MatchString("[[:upper:]/ !@#$%^&*()+~`=]", configName)
+	if matched {
+		msg := "ERROR: Config name '" + configName + "' must not contain upper case characters, spaces or special characters. Correct config name and rerun."
+		fmt.Println(msg)
+		return nil, errors.New(msg)
+	}
+
+	config_filename := string(strings.TrimRight(dir, "/") + "/" + config.Name + ".yml")
+	content, err := os.ReadFile(config_filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			fmt.Println("config file does not exist: " + config_filename)
+		}
+		return nil, err
+	}
+	baseConfig := &Config{}
+
+	if err := yaml.Unmarshal(content, baseConfig); err != nil {
+		return nil, err
+	}
+
+	if includeTemplates {
+		for _, t := range baseConfig.Templates {
+			if err := config.loadTemplate(templatesDir, t); err != nil {
+				return nil, err
+			}
+		}
+	}
+	if err := mergo.Merge(config, baseConfig, mergo.WithOverride); err != nil {
+		return nil, err
+	}
+	config.rawYaml = append(config.rawYaml, string(content[:]))
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range config.Labels {
+		val := strings.ReplaceAll(v, "{{config}}", config.Name)
+		config.Labels[k] = val
+	}
+
+	for k, v := range config.Env {
+		val := strings.ReplaceAll(v, "{{config}}", config.Name)
+		config.Env[k] = val
+	}
+
+	return config, nil
+}
+
+func (config *Config) Yaml() string {
+	return strings.Join(config.rawYaml, "_FILE_SEPERATOR_")
+}
+
+func (config *Config) Dockerfile(pupsArgs string, bakeEnv bool) string {
+	builder := strings.Builder{}
+	builder.WriteString("ARG dockerfile_from_image=" + config.Base_Image + "\n")
+	builder.WriteString("FROM ${dockerfile_from_image}\n")
+	builder.WriteString(config.DockerfileArgs() + "\n")
```

Comment:
I am not sure I understand this. The envs being declared in the YAML file is being declared as `ARG`in the Dockerfile. However, I don't see us using the `ARG`s in the Dockerfile which we generate here.

Reply:
`ARG` similar to `ENV` also gets passed through to the docker's environment, so if your YAML file has:

```
env:
  DISCOURSE_HOSTNAME: 'test.sample.com'
```

that will be set in the dockerfile as `ARG DISCOURSE_HOSTNAME=${DISCOURSE_HOSTNAME}`
on build, this sets the builder's shell environment, (ie: `export DISCOURSE_HOSTNAME=test.sample.com`) and run the docker build.

When building, this will bind the ARG DISCOURSE_HOSTNAME to be `test.sample.com`, and the build will see its environment set up as expected, as `$DISCOURSE_HOSTNAME=test.sample.com` for use during the pups runs.

(Sidenote, I'm passing env var values in the builder's shell environment as this is the only way I've found to pass multiline values.)

Reply:
> this sets the builder's shell environment, (ie: `export DISCOURSE_HOSTNAME=test.sample.com`) and run the docker build.

O this is something new I learnt. I thought the scope of `ARG` is only applied within the `Dockerfile`.